cpp/src/cluster/detail/kmeans.cuh

```diff
@@ -0,0 +1,1269 @@
+ * centroids as the initial cluster centers
+ *                              [out] Otherwise, generated centroids from the
+ * kmeans algorithm is stored at the address pointed by 'centroids'.
+ * @param[out]    inertia       Sum of squared distances of samples to their
+ * closest cluster center.
+ * @param[out]    n_iter        Number of iterations run.
+ */
+template <typename DataT, typename IndexT>
+void kmeans_fit(raft::resources const& handle,
+                const cuvs::cluster::kmeans::params& pams,
+                raft::device_matrix_view<const DataT, IndexT> X,
+                std::optional<raft::device_vector_view<const DataT, IndexT>> sample_weight,
+                raft::device_matrix_view<DataT, IndexT> centroids,
+                raft::host_scalar_view<DataT> inertia,
+                raft::host_scalar_view<IndexT> n_iter)
+{
+  raft::common::nvtx::range<raft::common::nvtx::domain::raft> fun_scope("kmeans_fit");
+  auto n_samples      = X.extent(0);
+  auto n_features     = X.extent(1);
+  auto n_clusters     = pams.n_clusters;
+  cudaStream_t stream = raft::resource::get_cuda_stream(handle);
+  // Check that parameters are valid
+  if (sample_weight.has_value())
+    RAFT_EXPECTS(sample_weight.value().extent(0) == n_samples,
+                 "invalid parameter (sample_weight!=n_samples)");
+  RAFT_EXPECTS(n_clusters > 0, "invalid parameter (n_clusters<=0)");
+  RAFT_EXPECTS(pams.tol > 0, "invalid parameter (tol<=0)");
+  RAFT_EXPECTS(pams.oversampling_factor >= 0, "invalid parameter (oversampling_factor<0)");
+  RAFT_EXPECTS((int)centroids.extent(0) == pams.n_clusters,
+               "invalid parameter (centroids.extent(0) != n_clusters)");
+  RAFT_EXPECTS(centroids.extent(1) == n_features,
+               "invalid parameter (centroids.extent(1) != n_features)");
+
+  // Display a message if the batch size is smaller than n_samples but will be ignored
+  if (pams.batch_samples < (int)n_samples &&
+      (pams.metric == cuvs::distance::DistanceType::L2Expanded ||
+       pams.metric == cuvs::distance::DistanceType::L2SqrtExpanded)) {
+    RAFT_LOG_DEBUG(
+      "batch_samples=%d was passed, but batch_samples=%d will be used (reason: "
+      "batch_samples has no impact on the memory footprint when FusedL2NN can be used)",
+      pams.batch_samples,
+      (int)n_samples);
+  }
+  // Display a message if batch_centroids is set and a fusedL2NN-compatible metric is used
+  if (pams.batch_centroids != 0 && pams.batch_centroids != pams.n_clusters &&
+      (pams.metric == cuvs::distance::DistanceType::L2Expanded ||
+       pams.metric == cuvs::distance::DistanceType::L2SqrtExpanded)) {
+    RAFT_LOG_DEBUG(
+      "batch_centroids=%d was passed, but batch_centroids=%d will be used (reason: "
+      "batch_centroids has no impact on the memory footprint when FusedL2NN can be used)",
+      pams.batch_centroids,
+      pams.n_clusters);
+  }
+
+  //  raft::logger::get(RAFT_NAME).set_level(pams.verbosity);
```

Comment:
There are many more comments like this one. Should we keep and uncomment or remove?

Reply:
This was left over from when I was trying to figure out compile errors. Those have been fixed so I can uncomment.

Reply:
Oh I see. Do we have to use `RAFT_NAME` or are we able to initialize the logger to "cuvs"?

Reply:
For now we do. We need to get the logger issue figured out soon, but I think that's a lower priority. (Raft name is unfortunately what's hardcoded in raft/core/logger.hpp)

Reply:
Oh my bad. That's what I get for responding on my phone. Yeah we probably don't have to use raft name here, but unfortunately the way the logger.hpp is structure in raft it forces us to use RAFT_NAME everywhere else... thus changing it here doesn't really do much for us. I can change it in this file though just for the heck of it.

Reply:
So we need to do `RAFT_NAME = "cuvs"` for it to work?

Reply:
Or `CUVS_NAME = "raft"`?

Reply:
Unfortunately, today we need to have RAFT_NAME defined for the logger itself to work. We don't have time for a better solution for 24.06 (there's still a lot of work to do and we need to get the guts migrated first).